src/main/java/net/dv8tion/jda/internal/components/actionrow/ActionRowImpl.java

```
@@ -0,0 +1,366 @@
+
+        List<ActionRow> rows = new ArrayList<>();
+        // The current action row we are building
+        List<ActionRowChildComponentUnion> currentRow = null;
+        // The component types contained in that row (for now it can't have mixed types)
+        Component.Type type = null;
+
+        for (ActionRowChildComponentUnion current : components)
+        {
+            if (type != current.getType() || currentRow.size() == ActionRow.getMaxAllowed(type))
+            {
+                type = current.getType();
+                ActionRowImpl row = (ActionRowImpl) ActionRow.of(current);
+                currentRow = row.components;
+                rows.add(row);
+            }
+            else
+            {
+                currentRow.add(current);
+            }
+        }
+
+        return rows;
+    }
+
+    @Override
+    public int getUniqueId()
+    {
+        return uniqueId;
+    }
+
+    @Nonnull
+    @Override
+    public List<ActionRowChildComponentUnion> getComponents()
+    {
+        return components;
+    }
+
+    @Nonnull
+    @Override
+    public ActionRow replace(@Nonnull ComponentReplacer replacer)
+    {
+        Checks.notNull(replacer, "ComponentReplacer");
+
+        return ComponentsUtil.doReplace(
+                ActionRowChildComponent.class,
+                getComponents(),
+                replacer,
+                newComponents -> new ActionRowImpl(newComponents, uniqueId)
+        );
+    }
+
+    @Nonnull
+    @Override
+    public ActionRow withUniqueId(int uniqueId)
+    {
+        Checks.positive(uniqueId, "Unique ID");
+        return new ActionRowImpl(getComponents(), uniqueId);
+    }
+
+    @Nonnull
+    @Override
+    public ActionRow createCopy()
+    {
+        return new ActionRowImpl(components, uniqueId);
+    }
+
+    // TODO after removal, make this immutable starting from the constructor
```

This is still a TODO, I'm think about just removing the modifiable behavior

---

It would be nicer for migration purposes to let this be mutable until the deprecated stuff gets removed.

If we suddenly make it immutable, all `updateComponent` methods on `LayoutComponent` won't be usable, they will fail silently unless removed

---

Unless we bump the major version, I don't think that's a good idea